### ISSUE TITLE
[BUG] Dashboard deserialization error when querying jobs

### ISSUE THREAD

**Issue by junhuhdev** [jobrunr/jobrunr#81]
http://localhost:8000/api/jobs?state=SUCCEEDED&offset=0&limit=20&order=updatedAt:DESC

I'm getting

```
java.lang.IllegalArgumentException: Class not found: org.jobrunr.jobs.context.JobDashboardLogger$JobDashboardLogLines
	at org.jobrunr.utils.reflection.ReflectionUtils.toClass(ReflectionUtils.java:84)
	at org.jobrunr.utils.mapper.gson.ClassNameObjectTypeAdapter.read(ClassNameObjectTypeAdapter.java:53)
	at com.google.gson.internal.bind.TypeAdapterRuntimeTypeWrapper.read(TypeAdapterRuntimeTypeWrapper.java:41)
	at com.google.gson.internal.bind.MapTypeAdapterFactory$Adapter.read(MapTypeAdapterFactory.java:187)
	at com.google.gson.internal.bind.MapTypeAdapterFactory$Adapter.read(MapTypeAdapterFactory.java:145)
	at com.google.gson.TypeAdapter.fromJsonTree(TypeAdapter.java:285)
	at org.jobrunr.utils.mapper.gson.RuntimeClassNameTypeAdapterFactory$1.read(RuntimeClassNameTypeAdapterFactory.java:219)
	at com.google.gson.TypeAdapter$1.read(TypeAdapter.java:199)
	at com.google.gson.TypeAdapter.fromJsonTree(TypeAdapter.java:285)
	at org.jobrunr.utils.mapper.gson.RuntimeClassNameTypeAdapterFactory$1.read(RuntimeClassNameTypeAdapterFactory.java:219)
	at com.google.gson.TypeAdapter$1.read(TypeAdapter.java:199)
	at com.google.gson.TypeAdapter.fromJsonTree(TypeAdapter.java:285)
	at org.jobrunr.utils.mapper.gson.RuntimeClassNameTypeAdapterFactory$1.read(RuntimeClassNameTypeAdapterFactory.java:219)
	at com.google.gson.TypeAdapter$1.read(TypeAdapter.java:199)
	at com.google.gson.TypeAdapter.fromJsonTree(TypeAdapter.java:285)
	at org.jobrunr.utils.mapper.gson.RuntimeClassNameTypeAdapterFactory$1.read(RuntimeClassNameTypeAdapterFactory.java:211)
	at com.google.gson.TypeAdapter$1.read(TypeAdapter.java:199)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.read(ReflectiveTypeAdapterFactory.java:131)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:222)
	at com.google.gson.TypeAdapter.fromJsonTree(TypeAdapter.java:285)
	at org.jobrunr.utils.mapper.gson.RuntimeClassNameTypeAdapterFactory$1.read(RuntimeClassNameTypeAdapterFactory.java:219)
	at com.google.gson.TypeAdapter$1.read(TypeAdapter.java:199)
	at com.google.gson.TypeAdapter.fromJsonTree(TypeAdapter.java:285)
	at org.jobrunr.utils.mapper.gson.RuntimeClassNameTypeAdapterFactory$1.read(RuntimeClassNameTypeAdapterFactory.java:219)
	at com.google.gson.TypeAdapter$1.read(TypeAdapter.java:199)
	at com.google.gson.TypeAdapter.fromJsonTree(TypeAdapter.java:285)
	at org.jobrunr.utils.mapper.gson.RuntimeClassNameTypeAdapterFactory$1.read(RuntimeClassNameTypeAdapterFactory.java:219)
	at com.google.gson.TypeAdapter$1.read(TypeAdapter.java:199)
	at com.google.gson.TypeAdapter.fromJsonTree(TypeAdapter.java:285)
	at org.jobrunr.utils.mapper.gson.RuntimeClassNameTypeAdapterFactory$1.read(RuntimeClassNameTypeAdapterFactory.java:219)
	at com.google.gson.TypeAdapter$1.read(TypeAdapter.java:199)
	at com.google.gson.Gson.fromJson(Gson.java:932)
	at com.google.gson.Gson.fromJson(Gson.java:897)
	at com.google.gson.Gson.fromJson(Gson.java:846)
	at com.google.gson.Gson.fromJson(Gson.java:817)
	at com.tele2.orderflow.slws.core.newflow.config.jobrunr.JobRunrGsonMapper.deserialize(JobRunrGsonMapper.java:76)
	at org.jobrunr.jobs.mappers.JobMapper.deserializeJob(JobMapper.java:20)
	at com.tele2.orderflow.slws.core.newflow.config.jobrunr.JobRunrJobMapper.deserializeJob(JobRunrJobMapper.java:25)
	at org.jobrunr.storage.sql.common.JobTable.toJob(JobTable.java:226)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at org.jobrunr.storage.sql.common.db.SqlSpliterator.tryAdvance(SqlSpliterator.java:44)
	at java.base/java.util.Spliterator.forEachRemaining(Spliterator.java:326)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
	at org.jobrunr.storage.sql.common.JobTable.selectJobsByState(JobTable.java:144)
	at org.jobrunr.storage.sql.common.DefaultSqlStorageProvider.getJobs(DefaultSqlStorageProvider.java:132)
	at org.jobrunr.storage.sql.common.DefaultSqlStorageProvider.getJobPage(DefaultSqlStorageProvider.java:157)
	at org.jobrunr.storage.ThreadSafeStorageProvider.getJobPage(ThreadSafeStorageProvider.java:122)
	at org.jobrunr.dashboard.JobRunrApiHandler.lambda$findJobByState$4(JobRunrApiHandler.java:67)
	at org.jobrunr.dashboard.server.http.RestHttpHandler.processRequest(RestHttpHandler.java:79)
	at org.jobrunr.dashboard.server.http.RestHttpHandler.lambda$handle$0(RestHttpHandler.java:71)
	at java.base/java.util.Optional.ifPresent(Optional.java:183)
	at org.jobrunr.dashboard.server.http.RestHttpHandler.handle(RestHttpHandler.java:71)
	at jdk.httpserver/com.sun.net.httpserver.Filter$Chain.doFilter(Filter.java:77)
	at jdk.httpserver/sun.net.httpserver.AuthFilter.doFilter(AuthFilter.java:82)
	at jdk.httpserver/com.sun.net.httpserver.Filter$Chain.doFilter(Filter.java:80)
	at jdk.httpserver/sun.net.httpserver.ServerImpl$Exchange$LinkHandler.handle(ServerImpl.java:692)
	at jdk.httpserver/com.sun.net.httpserver.Filter$Chain.doFilter(Filter.java:77)
	at jdk.httpserver/sun.net.httpserver.ServerImpl$Exchange.run(ServerImpl.java:664)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1426)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177)
Caused by: java.lang.ClassNotFoundException: org.jobrunr.jobs.context.JobDashboardLogger$JobDashboardLogLines
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:583)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:521)
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:398)
	at org.jobrunr.utils.reflection.ReflectionUtils.loadClass(ReflectionUtils.java:94)
	at org.jobrunr.utils.reflection.ReflectionUtils.toClass(ReflectionUtils.java:82)
	... 67 more

```

It's working locally but not when im running as spring-boot executable jar, any hints?

**Comment by rdehuyss** [comment]
Hmmm, I wonder if the class is in the actual executable jar. Is it an option for you to check it?

I'll try to look into it this weekend or next Wednesday.

**Comment by junhuhdev** [comment]
Hi thank you for the quick reply. 

Hmm how do I check that?

When I check inside the app-exec.jar in /BOOT-INF/lib I see that both jobrunr-1.2.3.jar and jobrunr-spring-boot-starter-1.2.3.jar exists there.

**Comment by junhuhdev** [comment]
Btw the jobscheduler runs fine, but this error only occurs when I try to access the dashboard. But apparently not when running locally.. I've been scratching my head to find the root cause since I can't reproduce it locally..

**Comment by rdehuyss** [comment]
I think this will probably be a class loading issue... Not easy to troubleshoot. 

If you run the executable jar locally, I assume you have the same problem?

**Comment by junhuhdev** [comment]
Ok yes when I run the exec jar locally i can reproduce it. 

running java 11
```
java -jar xxx-exec.jar 
```

**Comment by junhuhdev** [comment]
Seems to be failing here 

```java
    public static Class<?> loadClass(String className) throws ClassNotFoundException {
        // why: support for quarkus:dev (see https://github.com/quarkusio/quarkus/issues/2809) and Spring Boot Live reload
        // Jackson uses this order also
        ClassLoader classLoader = currentThread().getContextClassLoader();
        if (classLoader != null) {
            return Class.forName(className, true, classLoader);
        }

        return Class.forName(className);
    }
```

fails when calling ```return Class.forName(className, true, classLoader);```

**Comment by junhuhdev** [comment]
Seems to have solved when I changed 

from

```java
    public static Class<?> loadClass(String className) throws ClassNotFoundException {
        // why: support for quarkus:dev (see https://github.com/quarkusio/quarkus/issues/2809) and Spring Boot Live reload
        // Jackson uses this order also
        ClassLoader classLoader = currentThread().getContextClassLoader();
        if (classLoader != null) {
            return Class.forName(className, true, classLoader);
        }

        return Class.forName(className);
    }
```

to

```java
	public static Class<?> loadClass(String className) throws ClassNotFoundException {
		// why: support for quarkus:dev (see https://github.com/quarkusio/quarkus/issues/2809) and Spring Boot Live reload
		// Jackson uses this order also
		try {
			return Class.forName(className);
		} catch (ClassNotFoundException e) {
			log.error("Failed class={}", className, e);
			return currentThread().getContextClassLoader().loadClass(className);
		}
	}
```

**Comment by rdehuyss** [comment]
Ok, thanks for the input! I personally would prefer the approach below as I'm afraid we will break quarkus/Spring Boot Live reload otherwise. Could you test and report back? If it's ok for you, I'll create a new release with it then.

```java
    public static Class<?> loadClass(String className) throws ClassNotFoundException {
        // why: support for quarkus:dev (see https://github.com/quarkusio/quarkus/issues/2809) and Spring Boot Live reload
        // Jackson uses this order also
        try {
            ClassLoader classLoader = currentThread().getContextClassLoader();
            if (classLoader != null) {
                return Class.forName(className, true, classLoader);
            }
        } catch (ClassNotFoundException e) {
            // support for Spring Boot Executable jar. See https://github.com/jobrunr/jobrunr/issues/81
        }
        return Class.forName(className);
    }
```

**Comment by junhuhdev** [comment]
Ah yes I suspected that! 

Yes I just tried the fix above and it works as well! Yes it's great for me, and I appreciate for the quick reply and fix!

I have still not understood the underlying issue, do you know why it behaves differently for exec jar? 

This is how it looks in my pom

```xml
            <plugin>
                <groupId>org.springframework.boot</groupId>
                <artifactId>spring-boot-maven-plugin</artifactId>
                <executions>
                    <execution>
                        <id>repackage</id>
                        <configuration>
                            <classifier>exec</classifier>
                        </configuration>
                    </execution>
                </executions>
                <configuration>
                </configuration>
            </plugin>

```

**Comment by rdehuyss** [comment]
Solved in release v1.2.4.
